README
WWW::Scraper::ISBN::Amazon_Driver
=================================

This distribution contains two Perl extensions for the ISBN search collection 
that is provide by WWW::Scraper::ISBN.

The drivers search for book information from the Amazon online catalog from
the US (amazon.com) and UK (amazon.co.uk).

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these modules:

  JSON
  WWW::Scraper::ISBN
  WWW::Scraper::ISBN::Record
  WWW::Scraper::ISBN::Driver
  WWW::Mechanize

For testing the module requires these modules:

  IO::File
  Test::More

For optional testing the module would like these modules:

  Test::CPAN::Meta
  Test::CPAN::Meta::JSON
  Test::Pod
  Test::Pod::Coverage

COPYRIGHT AND LICENSE

  Copyright (C) 2004-2014 Barbie for Miss Barbell Productions

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic Licence v2.


Changes
Revision history for WWW-Scraper-ISBN-Amazon_Driver
===================================================

0.40    2014-09-17
        - fixed license fields in META.json to be lists.
        - added new title/author pattern.

0.39    2014-08-14
        - further fixes for UK layout changes.

0.38    2014-08-04
        - fixed for UK website update.
        - fixed UK site testing.

0.37    2014-05-29
        - added conversion routines to obtain EAN13.
        - upgrade to WWW-Scraper-ISBN-1.00.
        - changed to regex test for weight data on US site.
        - fixed META tests to catch version differences.

0.36    2014-04-28
        - amazon.com updated HTML output.
        - fixed distribution name in META.

0.35    2014-04-05
        - reduce size of test data footprint.

0.34    2014-04-02
        - getting the size can still fail, if data not available.
        - extract parse functionality into separate method, to allow testing.
        - protect against missing values causing undef errors.
        - extended test suite.

0.33    2014-03-24
        - added Amazon_Driver.pm, to help with PAUSE permissions.

0.32    2013-11-08
        - amazon.com updated description for Perl Medic.

0.31    2013-10-19
        - added git repository links to metadata.

0.30    2013-08-11
        - added isbn-check.pl to examples.
        - License updated consistently to Artistic License v2.0.
        - META updates.
        - height, width, depth regex reworked, tests simplified.
        - image, thumb regex reworked, now uses JSON, tests simplified.
        - UK description regex updated.

0.29    2013-03-03
        - alternative regexes for AmazonUS titles (thanks to Franck Porcher).
        - fixed weight calculations.
        - fixed image pattern matches.
        - expose depth values.

0.28    2012-12-16
        - weight extract also accounts for pounds (lbs)
        - depth extract added
        - height width and depth order reset.
        - change file dates changed to meet W3CDTF standards.
        - updated test suite.

0.27    2012-05-13
        - additonal regexes for variations of the data returned by Amazon.com.
        - spelling fixes. (thanks to Florian Schlichting & Gregor Herrmann).
        - META tests no longer require specific versions.
        - added minimum perl version (5.006).
        - reworked Makefile.PL for clarity.

0.26    2011-05-06
        - fix to description pattern match.

0.25    2011-03-22
        - added further skip test if test books are withdrawn from the site.
        - upgraded to Test::CPAN::Meta::JSON for tests.
        - Amazon now have a broken ISBN-10 field. ISBN-10 != ISBN-13.
        - fixed Amazon UK pattern for link from search results page.

0.24    2010-10-29
        - amended skip test counts.

0.23    2010-10-07
        - amended pingtest() in network tests.

0.22    2010-10-06
        - fixed skip count.
        - provided alternate ping test methods.

0.21    2010-09-15
        - reworked test structure, so easier to add sample sets.
        - fixed return value of 'isbn' to be 'isbn13'.
        - added 'description' to the field set returned.

0.20    2010-09-14
        - protect against amending undefined strings.

0.19    2010-08-10
        - added better protection from network failures.
        - updated prerequisite requirements.

0.18    2010-08-02
        - added connection protection if client or server offline.
        - documentation updates.

0.17    2010-07-30
        - documentation fixes.

0.16    2010-07-29
        - updates to parses for both Amazon US & UK sites.
        - added fields (if known) for binding, pages, weight, width & height.
        - added ean13, which is the industry name for isbn-13.
        - more tests.
        - removal of Template::Extract in favour of plain regexes.
        - various package file changes (inc adding META.json).

0.15    2009-08-14
        - minor fixes to parsing of both Amazon US & UK sites.
        - added ping tests, to avoid failing when no network connection.
        - real failures produce diagnostics not skips now.
        - added more package tests.

0.14    2008-11-17
        - minor fix to parsing Amazon US site.

0.13    2008-06-27
        - added ISBN10 and ISBN13 keys to book data
        - added tests to check for ISBN10 and ISBN13 searches

0.12    2008-06-27
        - added checks to only run extra tests under AUTOMATED_TESTING
        - switched metatest to use Test-CPAN-Meta
        - upped META.ym spec version and link
        - fixed US/UK content extraction

0.11    2007-10-29
        - fixed further pattern matching in tests, thanks to testing by 
          David Cantrell.

0.10    2007-07-27
        - fixed image link pattern matching in tests, thanks to testing by 
          David Cantrell and Chris Williams.

0.09    2007-07-26
        - Amazon image links have changed again.
        - Amazon UK full size image link popup has changed again
        - more tests modified due to the change in text used on pages.

0.08    2007-03-13
        - Amazon appear to change their URLs on a daily basis, as such the URL
          tests have been updated ... again!

0.07    2007-03-01
        - Handwritten META.yml to conform to current specs.
        - Changed all instances of LICENCE to LICENSE.
        - Added META.yml test script.
        - Added LICENSE and Artistic files.
        - Added an examples directory. Nothing in it yet, except a request
          for interesting scripts.
        - Updated the pattern matching code as Amazon UK have changed 
          their page layouts.
        - Added better error checking and reporting.
        - Devel::Cover = 95.6%

0.06    2006-10-11
        - Fixed US/UK website link changes
        - Devel::Cover - 89.7%

0.05    2005-03-01
        - Data & Layout change on both US site
        - fixed URL test in UK link tests
        - changed from @ISA to 'use base'

0.04    2005-01-07

Makefile.PL
#!/usr/bin/perl

use strict;
use warnings;
use 5.006;

use ExtUtils::MakeMaker;

WriteMakefile(
    AUTHOR              => 'Barbie <barbie@cpan.org>',
    NAME                => 'WWW-Scraper-ISBN-Amazon_Driver',
    VERSION_FROM        => 'lib/WWW/Scraper/ISBN/AmazonUS_Driver.pm',
    ABSTRACT            => 'Search drivers for the Amazon online catalog',
    NO_META             => 1,
    PREREQ_PM           => {

        # prereqs
        'JSON'                          => '0',
        'WWW::Mechanize'                => '1.60',
        'WWW::Scraper::ISBN'            => '1.00',
        'WWW::Scraper::ISBN::Driver'    => '1.00',

        # build/test prereqs
        'Data::Dumper'                  => '0',
        'IO::File'                      => '0',
        'Test::More'                    => '0.70'

    }
);


lib/WWW/Scraper/ISBN/Amazon_Driver.pm
package WWW::Scraper::ISBN::Amazon_Driver;

use strict;
use warnings;

use vars qw($VERSION);
$VERSION = '0.40';

#--------------------------------------------------------------------------

=head1 NAME

WWW::Scraper::ISBN::Amazon_Driver - A collection of search drivers for Amazon.

=head1 SYNOPSIS

See parent class documentation (L<WWW::Scraper::ISBN::Driver>)

=head1 DESCRIPTION

Searches for book information from the UK & US Amazon online catalogs.

For specific usability of the drivers, please see the documentation for the
respective driver you require.

=head1 REQUIRES

Requires the following modules be installed:

L<WWW::Scraper::ISBN::Driver>,
L<WWW::Mechanize>,

=head1 SEE ALSO

L<WWW::Scraper::ISBN>,
L<WWW::Scraper::ISBN::Record>,
L<WWW::Scraper::ISBN::Driver>

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please send an email to barbie@cpan.org or submit a bug to the
RT system (http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-Scraper-ISBN-Amazon_Driver).
However, it would help greatly if you are able to pinpoint problems or even
supply a patch.

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me.

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT & LICENSE

  Copyright (C) 2004-2014 Barbie for Miss Barbell Productions

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic Licence v2.

=cut


lib/WWW/Scraper/ISBN/AmazonUK_Driver.pm
package WWW::Scraper::ISBN::AmazonUK_Driver;

use strict;
use warnings;

use vars qw($VERSION);
$VERSION = '0.40';

#--------------------------------------------------------------------------

=head1 NAME

WWW::Scraper::ISBN::AmazonUK_Driver - Search driver for Amazon.co.uk

=head1 SYNOPSIS

See parent class documentation (L<WWW::Scraper::ISBN::Driver>)

=head1 DESCRIPTION

Searches for book information from the (UK) Amazon online catalog.

=cut

#--------------------------------------------------------------------------

###########################################################################
# Inheritence

use base qw(WWW::Scraper::ISBN::Driver);

###########################################################################
# Modules

use WWW::Mechanize;
use JSON;

###########################################################################
# Variables

my $AMA_SEARCH = 'http://www.amazon.co.uk/s/ref=nb_sb_noss?url=search-alias%3Daps&x=18&y=16&field-keywords=';
my $AMA_URL = 'http://www.amazon.co.uk/[^/]+/dp/[\dX]+/ref=sr_1_1/';
my $IN2MM = 0.0393700787;   # number of inches in a millimetre (mm)
my $LB2G  = 0.00220462;     # number of pounds (lbs) in a gram
my $OZ2G  = 0.035274;       # number of ounces (oz) in a gram

#--------------------------------------------------------------------------

###########################################################################
# Public Interface

=head1 METHODS

=over 4

=item C<search()>

Creates a query string, then passes the appropriate form fields to the
Amazon (UK) server.

The returned page should be the correct catalog page for that ISBN. If not the
function returns zero and allows the next driver in the chain to have a go. If
a valid page is returned, the following fields are returned via the book hash:

  isbn          (now returns isbn13)
  isbn10
  isbn13
  ean13         (industry name)
  author
  title
  book_link
  thumb_link
  image_link
  pubdate
  publisher
  binding       (if known)
  pages         (if known)
  weight        (if known) (in grams)
  width         (if known) (in millimetres)
  height        (if known) (in millimetres)
  depth         (if known) (in millimetres)

The book_link, thumb_link and image_link refer back to the Amazon (UK) website.

=back

=cut

sub search {
	my $self = shift;
	my $isbn = shift;
	$self->found(0);
	$self->book(undef);

    # validate and convert into EAN13 format
    my $ean = $self->convert_to_ean13($isbn);
    return $self->handler("Invalid ISBN specified [$isbn]")
        if(!$ean || (length $isbn == 13 && $isbn ne $ean)
                 || (length $isbn == 10 && $isbn ne $self->convert_to_isbn10($ean)));

	my $mech = WWW::Mechanize->new();
    $mech->agent_alias( 'Linux Mozilla' );

    my $search = $AMA_SEARCH . $ean;

	eval { $mech->get( $search ) };
    return $self->handler("Amazon UK website appears to be unavailable.")
	    if($@ || !$mech->success() || !$mech->content());

	my $content = $mech->content();
#print STDERR "\n# content=[$content]\n";
    my ($link) = $content =~ m!($AMA_URL)!s;
	return $self->handler("Failed to find that book on Amazon UK website.")
	    unless($link);

	eval { $mech->get( $link ) };
    return $self->handler("Amazon UK website appears to be unavailable.")
	    if($@ || !$mech->success() || !$mech->content());

    return $self->_parse($mech);
}

sub _parse {
    my $self = shift;
    my $mech = shift;

	# The Book page
    my $html = $mech->content;
    my $data = {};

#print STDERR "\n# html=[$html]\n";

    my @size                            = $html =~ m!<li><b>\s*Product Dimensions:\s*</b>\s*([\d.]+) x ([\d.]+) x ([\d.]+) (cm)\s*</li>!si;
    @size                               = $html =~ m!<li><b>\s*Product Dimensions:\s*</b>\s*([\d.]+) x ([\d.]+) x ([\d.]+) (inches)\s*</li>!si unless(@size);
    if(@size) {
        my $type = pop @size;
        ($data->{depth},$data->{width},$data->{height}) = sort @size;
        if($type eq 'cm') {
            $data->{$_}  = int($data->{$_} * 10)  for(qw( height width depth ));
        } elsif($type eq 'inches') {
            $data->{$_}  = int($data->{$_} / $IN2MM)  for(qw( height width depth ));
        }
    }

    ($data->{binding},$data->{pages})   = $html =~ m!<li><b>(Paperback|Hardcover):</b>\s*([\d.]+)\s*pages</li>!si;
    ($data->{weight})                   = $html =~ m!<li><b>Shipping Weight:</b>\s*([\d.]+)\s*ounces</li>!si;
    ($data->{published})                = $html =~ m!<li><b>Publisher:</b>\s*(.*?)</li>!si;
    ($data->{isbn10})                   = $html =~ m!<li><b>ISBN-10:</b>\s*(.*?)</li>!si;
    ($data->{isbn13})                   = $html =~ m!<li><b>ISBN-13:</b>\s*(.*?)</li>!si;
    ($data->{content})                  = $html =~ m!<meta name="description" content="([^"]+)"!si;
    ($data->{description})              = $html =~ m!From the Back Cover</h3>\s*<div class="productDescriptionWrapper"\s*>\s*<P>(.*?)<div!si;
    ($data->{description})              = $html =~ m!<div id="bookDescription_feature_div"[^>]*>.*?<noscript>(.*?)</noscript!si unless($data->{description});

    $data->{weight} = int($data->{weight} / $OZ2G)  if($data->{weight});

    if($data->{description}) {
        $data->{description} =~ s!<[^>]+>!!g;
        $data->{description} =~ s! +! !g;
    }

    # The images
    my ($json) = $html =~ /var colorImages = ([^;]+);/si;
    if($json) {
        my $code = decode_json($json);
        my @order = grep {$_} $code->{initial}[0]{thumb}, $code->{initial}[0]{landing}, @{$code->{initial}[0]{main}}, $code->{initial}[0]{large};
        $data->{thumb_link} = $order[0]     if(@order);
        $data->{image_link} = $order[-1]    if(@order);

#use Data::Dumper;
#print STDERR "\n# code=[".Dumper($code)."]\n";
    } else {
        my ($code) = $html =~ /'imageGalleryData'\s*:\s*([^;]+);/si;
        if($code) {
            ($data->{thumb_link}) = $code =~ /"thumbUrl":\s*"([^+]+)"/;
            ($data->{image_link}) = $code =~ /"mainUrl":\s*"([^+]+)"/;
        }
#use Data::Dumper;
#print STDERR "\n# code=[".Dumper($code)."]\n";
    }


#    {\"initial\":[{\"large\":\"http://ecx.images-amazon.com/images/I/31cLTIXHKgL.jpg\",\"landing\":[\"http://ecx.images-amazon.com/images/I/31cLTIXHKgL._SY300_.jpg\"],\"thumb\":\"http://ecx.images-amazon.com/images/I/31cLTIXHKgL._SS40_.jpg\",\"main\":[\"http://ecx.images-amazon.com/images/I/31cLTIXHKgL._SX342_.jpg\",\"http://ecx.images-amazon.com/images/I/31cLTIXHKgL._SX385_.jpg\"]}]};

    if($data->{content}) {
        $data->{content} =~ s/Amazon\.co\.uk.*?://i;
        $data->{content} =~ s/: Books.*//i;
        ($data->{title},$data->{author}) = split(/\s+by\s+/,$data->{content});
        $data->{title}  =~ s/^Buy\s+//;
        $data->{author} =~ s/\s*\(.*//;
    }

    ($data->{publisher},$data->{pubdate}) = ($data->{published} =~ /\s*(.*?)(?:;.*?)?\s+\((.*?)\)/) if($data->{published});
    $data->{isbn10}  =~ s/[^\dX]+//g    if($data->{isbn10});
    $data->{isbn13}  =~ s/\D+//g        if($data->{isbn13});
	$data->{pubdate} =~ s/^.*?\(//      if($data->{pubdate});

	return $self->handler("Could not extract data from Amazon UK result page.")
		unless(defined $data->{isbn13});

    # trim top and tail
	foreach (keys %$data) { next unless(defined $data->{$_});$data->{$_} =~ s/^\s+//;$data->{$_} =~ s/\s+$//; }

#use Data::Dumper;
#print STDERR "\n# data=[".Dumper($data)."]\n";

	my $bk = {
		'ean13'		    => $data->{isbn13},
		'isbn13'		=> $data->{isbn13},
		'isbn10'		=> $data->{isbn10},
		'isbn'			=> $data->{isbn13},
		'author'		=> $data->{author},
		'title'			=> $data->{title},
		'image_link'	=> $data->{image_link},
		'thumb_link'	=> $data->{thumb_link},
		'publisher'		=> $data->{publisher},
		'pubdate'		=> $data->{pubdate},
		'book_link'		=> $mech->uri(),
		'content'		=> $data->{content},
		'binding'	    => $data->{binding},
		'pages'		    => $data->{pages},
		'weight'		=> $data->{weight},
		'width'		    => $data->{width},
		'height'		=> $data->{height},
        'depth'         => $data->{depth},
		'description'	=> $data->{description},
        'html'          => $html
	};
	$self->book($bk);
	$self->found(1);
	return $self->book;
}

q{currently reading: 'Torn Apart: The Life of Ian Curtis' by Mick Middles and Lindsay Reade};

__END__

=head1 REQUIRES

Requires the following modules be installed:

L<WWW::Scraper::ISBN::Driver>,
L<WWW::Mechanize>,

=head1 SEE ALSO

L<WWW::Scraper::ISBN>,
L<WWW::Scraper::ISBN::Record>,
L<WWW::Scraper::ISBN::Driver>

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please send an email to barbie@cpan.org or submit a bug to the
RT system (http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-Scraper-ISBN-Amazon_Driver).
However, it would help greatly if you are able to pinpoint problems or even
supply a patch.

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me.

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT & LICENSE

  Copyright (C) 2004-2014 Barbie for Miss Barbell Productions

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic Licence v2.

=cut


lib/WWW/Scraper/ISBN/AmazonUS_Driver.pm
package WWW::Scraper::ISBN::AmazonUS_Driver;

use strict;
use warnings;

use vars qw($VERSION);
$VERSION = '0.40';

#--------------------------------------------------------------------------

=head1 NAME

WWW::Scraper::ISBN::AmazonUS_Driver - Search driver for Amazon.com

=head1 SYNOPSIS

See parent class documentation (L<WWW::Scraper::ISBN::Driver>)

=head1 DESCRIPTION

Searches for book information from the (US) Amazon online catalog.

=cut

#--------------------------------------------------------------------------

###########################################################################
# Inheritence

use base qw(WWW::Scraper::ISBN::Driver);

###########################################################################
# Modules

use WWW::Mechanize;
use JSON;

###########################################################################
# Variables

my $AMA_SEARCH = 'http://www.amazon.com/s/ref=nb_sb_noss?url=search-alias=us-stripbooks-tree&field-keywords=%s';
my $AMA_URL    = 'http://www.amazon.com/[^/]+/dp/[\dX]+/ref=sr_1_1.*?sr=1-1';
my $IN2MM = 0.0393700787;   # number of inches in a millimetre (mm)
my $LB2G  = 0.00220462;     # number of pounds (lbs) in a gram
my $OZ2G  = 0.035274;       # number of ounces (oz) in a gram

#--------------------------------------------------------------------------

###########################################################################
# Public Interface

=head1 METHODS

=over 4

=item C<search()>

Creates a query string, then passes the appropriate form fields to the
Amazon (US) server.

The returned page should be the correct catalog page for that ISBN. If not the
function returns zero and allows the next driver in the chain to have a go. If
a valid page is returned, the following fields are returned via the book hash:

  isbn          (now returns isbn13)
  isbn10
  isbn13
  ean13         (industry name)
  author
  title
  book_link
  thumb_link
  image_link
  pubdate
  publisher
  binding       (if known)
  pages         (if known)
  weight        (if known) (in grams)
  width         (if known) (in millimetres)
  height        (if known) (in millimetres)
  depth         (if known) (in millimetres)

The book_link, thumb_link and image_link refer back to the Amazon (US) website.

=back

=cut

sub search {
    my $self = shift;
    my $isbn = shift;
    $self->found(0);
    $self->book(undef);

    # validate and convert into EAN13 format
    my $ean = $self->convert_to_ean13($isbn);
    return $self->handler("Invalid ISBN specified [$isbn]")
        if(!$ean || (length $isbn == 13 && $isbn ne $ean)
                 || (length $isbn == 10 && $isbn ne $self->convert_to_isbn10($ean)));

    my $mech = WWW::Mechanize->new();
    $mech->agent_alias( 'Linux Mozilla' );

    my $search = sprintf $AMA_SEARCH, $ean;

    eval { $mech->get( $search ) };
    return $self->handler("Amazon US website appears to be unavailable.")
        if($@ || !$mech->success() || !$mech->content());

    my $content = $mech->content();
    my ($link) = $content =~ m!($AMA_URL)!s;
    return $self->handler("Failed to find that book on Amazon US website.")
        unless($link);

    eval { $mech->get( $link ) };
    return $self->handler("Amazon US website appears to be unavailable.")
        if($@ || !$mech->success() || !$mech->content());

    return $self->_parse($mech);
}

sub _parse {
    my $self = shift;
    my $mech = shift;

    # The Book page
    my $html = $mech->content;
    my $data = {};

#print STDERR "\n# html=[$html]\n";

    my @patterns = (
        qr/<meta name="description" content="(.*?) \[(.*?)\] on Amazon.com/si,
        qr/<meta name="description" content="(.*?): (.*?): (\d+): Amazon.com: Books/si,
        qr/<meta name="description" content="(?:Amazon.com:)?\s*(.*?).\d+.:\s+([^:]+): Books/si,
        qr/<meta name="description" content="(?:Amazon.com: Books: )?\s*(.*?)(?:\s+by|,)\s+(.*)/si,
        qr/<meta name="description" content="(?:Amazon.com:)?\s*(.*?)(?:\s+by|,|:)\s+([^:]+): Books/si
    );

    for my $pattern (@patterns) {
        my ($title,$author) = $html =~ $pattern;
        $data->{title}  ||= $title;
        $data->{author} ||= $author;
#print STDERR "\n# title=[$data->{title}], author=[$data->{author}] pattern=[$pattern]\n";

        last    if($data->{title} && $data->{author});
    }

    ($data->{binding},$data->{pages})   = $html =~ m!<li><b>(Paperback|Hardcover):</b>\s*([\d.]+)\s*pages</li>!si;
    ($data->{published})                = $html =~ m!<li><b>Publisher:</b>\s*(.*?)</li>!si;
    ($data->{isbn10})                   = $html =~ m!<li><b>ISBN-10:</b>\s*(.*?)</li>!si;
    ($data->{isbn13})                   = $html =~ m!<li><b>ISBN-13:</b>\s*(.*?)</li>!si;
    ($data->{content})                  = $html =~ m!<meta name="description" content="([^"]+)"!si;

    @patterns = (
        qr!<h2>Book Description</h2>.*?<div id="postBodyPS"[^>]+>\s*<div[^>]*>\s*<p>\s*(.*?)\s*</p>\s*</div!si,
        qr!<h2>Book Description</h2>.*?<div id="postBodyPS"[^>]+>\s*<div[^>]*>\s*(.*?)\s*</div!si,
        qr!<h3 class="productDescriptionSource">(?:Product Description|From the Back Cover)</h3>\s*<div class="productDescriptionWrapper">\s*<p>([^<]+)!si,
        qr!<h3 class="productDescriptionSource">(?:Product Description|From the Back Cover)</h3>\s*<div class="productDescriptionWrapper">\s*(.*?)<div!si,
        qr!<div id="bookDescription_feature_div"[^>]+>\s*<script[^>]*>.*?</script>\s*<noscript>(.*?)</noscript>!si
    );

    for my $pattern (@patterns) {
        my ($desc) = $html =~ $pattern;
        $desc =~ s/\s+$//s   if($desc);
        $data->{description}  ||= $desc;

        last    if($data->{description});
    }

    for my $key (qw(description)) {
        next unless($data->{$key});
        $data->{$key} =~ s/<[^>]*>//gs;
    }

    # amazon use both ounces and pounds
    my $weight;
    ($data->{weight},$weight)                   = $html =~ m!<li><b>Shipping Weight:</b>\s*([\d.]+)\s*(ounces|pounds)!si;
    $data->{weight} = int($data->{weight} / $OZ2G)  if($data->{weight} && $weight eq 'ounces');
    $data->{weight} = int($data->{weight} / $LB2G)  if($data->{weight} && $weight eq 'pounds');

    # amazon change this regularly
    my @size                            = $html =~ m!<li><b>\s*Product Dimensions:\s*</b>\s*([\d.]+) x ([\d.]+) x ([\d.]+) (cm)\s*</li>!si;
    @size                               = $html =~ m!<li><b>\s*Product Dimensions:\s*</b>\s*([\d.]+) x ([\d.]+) x ([\d.]+) (inches)\s*</li>!si unless(@size);
    if(@size) {
        my $type = pop @size;
        ($data->{depth},$data->{width},$data->{height}) = sort @size;
        if($type eq 'cm') {
            $data->{$_}  = int($data->{$_} * 10)  for(qw( height width depth ));
        } elsif($type eq 'inches') {
            $data->{$_}  = int($data->{$_} / $IN2MM)  for(qw( height width depth ));
        }
    }

    # The images
    my ($json) = $html =~ /var colorImages = ([^;]+);/si;
    if($json) {
        my $code = decode_json($json);
        my @order = grep {$_} $code->{initial}[0]{thumb}, $code->{initial}[0]{landing}, @{$code->{initial}[0]{main}}, $code->{initial}[0]{large};
        $data->{thumb_link} = $order[0]     if(@order);
        $data->{image_link} = $order[-1]    if(@order);
    } else {
        ($data->{thumb_link}) = $html =~ m!imageGalleryData.*?thumbUrl":"([^"]+)"!;
        ($data->{image_link}) = $html =~ m!imageGalleryData.*?mainUrl":"([^"]+)"!;
    }

    ($data->{publisher},$data->{pubdate}) = ($data->{published} =~ /\s*(.*?)(?:;.*?)?\s+\((.*?)\)/) if($data->{published});
    $data->{isbn10} =~ s/[^\dX]+//g if($data->{isbn10});
    $data->{isbn13} =~ s/\D+//g     if($data->{isbn13});


    return $self->handler("Could not extract data from Amazon US result page.")
        unless(defined $data->{isbn13});

    # trim top and tail
    foreach (keys %$data) { next unless(defined $data->{$_});$data->{$_} =~ s/^\s+//;$data->{$_} =~ s/\s+$//; }

    my $bk = {
        'ean13'         => $data->{isbn13},
        'isbn13'        => $data->{isbn13},
        'isbn10'        => $data->{isbn10},
        'isbn'          => $data->{isbn13},
        'author'        => $data->{author},
        'title'         => $data->{title},
        'image_link'    => $data->{image_link},
        'thumb_link'    => $data->{thumb_link},
        'publisher'     => $data->{publisher},
        'pubdate'       => $data->{pubdate},
        'book_link'     => $mech->uri(),
        'content'       => $data->{content},
        'binding'       => $data->{binding},
        'pages'         => $data->{pages},
        'weight'        => $data->{weight},
        'width'         => $data->{width},
        'height'        => $data->{height},
        'depth'         => $data->{depth},
        'description'   => $data->{description},
        'html'          => $html
    };
    $self->book($bk);
    $self->found(1);
    return $self->book;
}

q{currently reading: 'Soul Music' by Terry Pratchett};

__END__

=head1 REQUIRES

Requires the following modules be installed:

L<WWW::Scraper::ISBN::Driver>,
L<WWW::Mechanize>,

=head1 SEE ALSO

L<WWW::Scraper::ISBN>,
L<WWW::Scraper::ISBN::Record>,
L<WWW::Scraper::ISBN::Driver>

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please send an email to barbie@cpan.org or submit a bug to the
RT system (http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-Scraper-ISBN-Amazon_Driver).
However, it would help greatly if you are able to pinpoint problems or even
supply a patch.

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me.

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT & LICENSE

  Copyright (C) 2004-2014 Barbie for Miss Barbell Productions

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic Licence v2.

=cut
